### ISSUE TITLE
FTPS support

### ISSUE THREAD

**Issue by marcandrews** [kooohei/brackets-synapse#64]
Please add FTPS support

**Comment by marcandrews** [comment]
Seems you're using [node-ftp](https://github.com/mscdex/node-ftp), which supports FTPS. You could try [node-ftps](https://github.com/Atinux/node-ftps), which supports FTP, FTPS and SFTP.

**Comment by kooohei** [comment]
yea I know.
I was try that and node-ftp, jsftp.
node-ftps is use standalone, should install "lftp" on environment.
i.e just a wrapper.
jsftp able to use ftps.
if I add to ftps function to synapse. I'll use that.

**Comment by marcandrews** [comment]
Can jsftp do SFTP too? If it can, you should definitely use that. I can help convert some functions in SynapseDomain.js from node-ftp to jsftp.

**Comment by kooohei** [comment]
could not sftp, jsftp can FTP[S]

**Comment by marcandrews** [comment]
I'd really like FTP**S** support.

I was really hoping there was a module that would give you all three FTP, SFTP and FTPS, and thus decrease the amount functions required in Synapse Domain.is, but it looks like there isn't one.

**Comment by marcandrews** [comment]
Wait, node-ftp supports FTPS so it should be easy to add FTPS support without adding or changing to another module.

**Comment by marcandrews** [comment]
OK, so I just tested FTPS connections with node-ftp. All you have to do is add ```secure: true``` to your connect object config and it should work. Ideally, you'd probably want to add some UI to choose between the available options for ```secure``` (i.e. ```true```, ```'control'``` or ```'implicit'```).

**Comment by marcandrews** [comment]
OK so I added FTPS support. Which branch should I fork so that I suggest a pull request?

**Comment by kooohei** [comment]
I wake now. oh thanks! I will confirm it from now. 
ok plz fork from master :)

**Comment by kooohei** [comment]
so FTPS different to FTP is only port number and secure option via node-ftp right?

**Comment by marcandrews** [comment]
Port number is the same ... so it's just the ```secure``` option.

I forked your master and made some changes, but when I tried to load it into Brackets, it failed. So made changes to the version on Brackets > Extensions and it worked. Is your master on GitHub in sync and and uptodate with the version on Extensions?

**Comment by kooohei** [comment]
sorry plz try again from master sync.

**Comment by marcandrews** [comment]
Yeah, I don't know what's wrong with your repo. When I download the ZIP of your master, and drag it to Brackets > Extensions, the installation always fails, so I can't test it.

**Comment by kooohei** [comment]
do you do npm install?

**Comment by marcandrews** [comment]
How exactly do I do that? I'm running Brackets on Windows 10.

I can usually just download the ZIP from GitHub and drag it to to Brackets > Extensions and it installs fine, but with your repo, I can't do that. Any ideas?

**Comment by kooohei** [comment]
Oh realy?
I could not.
Ok what you see in the brackets-synapse folder?

**Comment by marcandrews** [comment]
It's exactly the same as your master

**Comment by kooohei** [comment]
If you can see the "node" directory then make sure is exist node_modules directly into that.

**Comment by marcandrews** [comment]
It's in your master, so it's in the ZIP I download from GitHub

**Comment by kooohei** [comment]
Ah I see.
First you should fork my master repos by git.
And nodejs and mom install your pc.
Then execute the npm install inside of the forked dirs

**Comment by kooohei** [comment]
Would not pull request from download files..

**Comment by kooohei** [comment]
and Your modified FTPS is included modes all of them?

**Comment by marcandrews** [comment]
I'm on Windows so I can't easily run npm install. Anyway, it seems like your repo on GitHub is missing crypto-js, font-awesome, homedir, q, and ssh2 from node/node_modules ... probably because I can't run npm install ... but everything else looks the same.

So I'll just update my fork with my changes and hope for the best :-)

**Comment by marcandrews** [comment]
OK pull request created

**Comment by kooohei** [comment]
I had a UI, already created.


I create new branche, can you pull request from that?
and plz request with the only change.
but I could not confirm, where Implict mode?

**Comment by marcandrews** [comment]
OK I created a pull request for your ftps branch.

I can't really make a pull request with only one change because they are all related to adding FTPS support.
- https://github.com/marcandrews/brackets-synapse/commit/52c44c79dd8cad302a04b052ff069f9ff1d9d612, https://github.com/marcandrews/brackets-synapse/commit/7a64b4e5752ce46595c5f46b6b995a51054fad30 and https://github.com/marcandrews/brackets-synapse/commit/74e3abf1e55274e23a93120e81d0be49f43dd1b7 are the changes I made before I realized that I couldn't test your repo directly in Brackets
- https://github.com/marcandrews/brackets-synapse/commit/46b4b2e1fad9c8f85db7e45da6d88d9ea63cf349 merges the changes you made to your master after I forked it into my forked repo
- https://github.com/marcandrews/brackets-synapse/commit/668bfd1179e0ae7472b6870f102ce4443c807865 are some additional changes to fix issues that came up once I was able to test

**Comment by kooohei** [comment]
no I want your pull request.
I think that it may be updated until the release branch.

**Comment by kooohei** [comment]
I only minimal functionality I want included.
because When something is not enough that is requested anytime soon to me ><

**Comment by kooohei** [comment]
I already said UI created.
confirm your code, now I don't see differences.
it is not a complete FTPS is only waring a secure option.
how selected ssl and tls?

**Comment by marcandrews** [comment]
But that is my pull request ... it will merge all my changes with your repo. Your repo already has 46b4b2e so those changes won't be made again obviously ... so what you'll end up with is my changes added to your repo.

I already created the UI and most of the back end in my repo so it'll be hard for me to revert and integrate my backend with your frontend.

Also, you don't need anything else besides secure = true or control or explicit. To my knowledge, you don't have to choose between SSL and TLS for FTPS.

**Comment by marcandrews** [comment]
Sorry about the white-space changes :-(

**Comment by kooohei** [comment]
I'm so sorry for the late.
I have confused my server setting for the FTPS
I understand to see node-ftp wiki
and I make sure your code done. You right, I'll change to only Now I'm out, finally I will confirm to data on connection actualy.

realy sorry for too many request, but you know? We must nervous to related things of the security.

so thanks a lot, and nice work!

**Comment by marcandrews** [comment]
No worries. I'm just happy I can add to this amazing project.

Let me know if you need help testing.

**Comment by kooohei** [comment]
Hi, I have been try implicit mode, but that was not encrypted handshake.
and Now, Your solution (process.env.NODE_REJECT_UNAUTHORIZED) is really bad ideas in some way.
but I confirm this trick at old discussion on any place.
the problem is defferent too much, because NodeJS significanttly changed stream and tls class.
even worse, The node.exe of the Brackets is a little old.
plz retry add your code to the reinstall synapse from the Extension Manager of the Brackets.
that would not connected to server via the implicit mode or it is not the implicit mode actual if it connected.

This problem is difficult, because the compatibility of injection is required part is the external library.
node-ftp will probably migrate to the new tls.socket system.
owever NodeJs of Brackets is probably remain old.
Do not depend on the Nodejs, or it sounds like no choice but to wait for the timing.

**Comment by marcandrews** [comment]
Sorry, I had some issues with Brackets but it seems everything is working again.

When I tested FTPS, both explicit and implicit connections worked; I will reinstall the FTPS branch and test again to confirm.

When you were doing your own tests with implicit FTPS, was your port 990 forwarded/open?

**Comment by kooohei** [comment]
sure and I confirm my packet with wireshark, but that is not encrypted.
does that fallback to explicit when your tested?

**Comment by kooohei** [comment]
and I think that the error handling technique is fatal.
Basically, The client should be self cert is deprecated.
Now, that could not the CA trust error. right?

**Comment by viktorix** [comment]
What's the status on this? 

Thanks.

**Comment by dhaupin** [comment]
We also need FTPS in brackets, as none of the extensions seem to handle that -- is this being applied at some point?
